pkg/scheduler/framework/genericpreemption/preemption.go

```diff
@@ -0,0 +1,514 @@
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/metrics"
+	"k8s.io/kubernetes/pkg/scheduler/util"
+)
+
+// Candidate represents a nominated node on which the preemptor can be scheduled,
+// along with the list of victims that should be evicted for the preemptor to fit the node.
+type Candidate interface {
+	// Victims wraps a list of to-be-preempted Pods and the number of PDB violation.
+	Victims() *extenderv1.Victims
+	// Name returns the target node name where the preemptor gets nominated to run.
+	Name() string
+}
+
+type candidate struct {
+	victims *extenderv1.Victims
+	name    string
+}
+
+// Victims returns s.victims.
+func (s *candidate) Victims() *extenderv1.Victims {
+	return s.victims
+}
+
+// Name returns s.name.
+func (s *candidate) Name() string {
+	return s.name
+}
+
+// PreemptionInteface is expected to be implemented by different preemption plugins as all those member
+// methods might have different behavior compared with the default preemption.
+type PreemptionInteface interface {
+	// GetOffsetAndNumCandidates chooses a random offset and calculates the number of candidates that should be
+	// shortlisted for dry running preemption.
+	GetOffsetAndNumCandidates(nodes int32) (int32, int32)
+	// CandidatesToVictimsMap builds a map from the target node to a list of to-be-preempted Pods and the number of PDB violation.
+	CandidatesToVictimsMap(candidates []Candidate) map[string]*extenderv1.Victims
+	// DryRunPreemption simulates Preemption logic on target node, returns preemption candidates and a map indicating filtered nodes statuses.
+	DryRunPreemption(context.Context, framework.Handle, *framework.CycleState, *v1.Pod, []*framework.NodeInfo, []*policy.PodDisruptionBudget, int32, int32) ([]Candidate, framework.NodeToStatusMap)
+}
+
+type PreemptionEvaluator struct {
+	Name      string
+	Handler   framework.Handle
+	PodLister corelisters.PodLister
+	PdbLister policylisters.PodDisruptionBudgetLister
+	PreemptionInteface
+}
+
+func (ev *PreemptionEvaluator) Preempt(ctx context.Context, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, *framework.Status) {
+	cs := ev.Handler.ClientSet()
+	nodeLister := ev.Handler.SnapshotSharedLister().NodeInfos()
+
+	// 0) Fetch the latest version of <pod>.
+	// It's safe to directly fetch pod here. Because the informer cache has already been
+	// initialized when creating the Scheduler obj, i.e., factory.go#MakeDefaultErrorFunc().
+	// However, tests may need to manually initialize the shared pod informer.
+	podNamespace, podName := pod.Namespace, pod.Name
+	pod, err := ev.PodLister.Pods(pod.Namespace).Get(pod.Name)
+	if err != nil {
+		klog.ErrorS(err, "getting the updated preemptor pod object", "pod", klog.KRef(podNamespace, podName))
+		return "", framework.AsStatus(err)
+	}
+
+	// 1) Ensure the preemptor is eligible to preempt other pods.
+	if !PodEligibleToPreemptOthers(pod, nodeLister, m[pod.Status.NominatedNodeName]) {
```

Comment:
>It's actually not that accurate - we should verify the terminating pods are exactly preemption victims; otherwise a possible case is: the victims have been physically deleted, and the terminating pods are due to regular deletion.

Good points! let's see whether we can improve this logic later. And in this PR, I have move the function into `Interface`.